File: HeadFirstDesignPatternsOC/DecoratorPattern/Soy.m
//
//  Soy.m
//  DecoratorPattern
//
//  Created by 陳峻琦 on 27/5/2022.
//

#import "Soy.h"

@implementation Soy

- (NSString *)description {
    return [NSString stringWithFormat:@"%@, Soy", self.beverage];
}

- (double)cost {
    return 0.99 + [self.beverage cost];
}

@end

File: HeadFirstDesignPatternsJava/HeadFirst/src/DecoratorPattern/Soy.java
package DecoratorPattern;

public class Soy extends CondimentDecorator {
    Beverage beverage;

    public Soy(Beverage beverage) {
        this.beverage = beverage;
    }

    @Override
    public String gerDescription() {
        return beverage.gerDescription() + "Soy";
    }

    @Override
    public double cost() {
        return 0.99 + beverage.cost();
    }
    
}
